24.0.0.7 GA release post (#3871)
OpenLiberty/blogs: posts/2024-07-16-24.0.0.7.adoc

// Contact/Reviewer: tonyreigns
// // // // // // // //
[#mpm]
== Enhanced histogram and timer metrics in MicroProfile 3.0 and 4.0

This release introduces MicroProfile Config properties for MicroProfile 3.0 and 4.0 that are used for configuring the statistics that are tracked and outputted by the histogram and timer metrics. These changes are already available in MicroProfile Metrics 5.1.

In previous MicroProfile Metrics 3.0 and 4.0 releases, histogram and timer metrics tracked only the following values:

* Min/max recorded values
* The sum of all values
* The count of the recorded values
* A static set of percentiles for the 50th, 75th, 95th, 98th, 99th, and 99.9th percentile.

These values are output to the `/metrics` endpoint in Prometheus format.

The new properties can define a custom set of percentiles as well as custom set of histogram buckets for the histogram and timer metrics. Other new configuration properties can enable a default set of histogram buckets, including properties that define an upper and lower bound for the bucket set.

With these properties, you can define a semicolon-separated list of value definitions that use the following syntax:

----

> <metric name>=<value-1>[,<value-2>…<value-n>]

Reviewer: ```suggestion
<metric name>=<value-1>[,<value-2>...<value-n>]
```
Perhaps a nit, but if we're using a code snippet I would prefer to have the ellipses be three `.` characters rather than the single `…` U+2026 Unicode character. That might be a personal preference, but the U+2026 character looks odd in monospace text.

----

Some properties can accept multiple values for a given metric name, while others can accept only a single value.

> You can use an asterisk ( *) as a wildcard at the end of the metric name.

Reviewer: ```suggestion
You can use an asterisk (`*`) as a wildcard at the end of the metric name.
```

[cols="2*"]
|===
| Property  | Description
| mp.metrics.distribution.percentiles | Defines a custom set of percentiles for matching Histogram and Timer metrics to track and output. Accepts a set of integer and decimal values for a metric name pairing. Can be used to disable percentile output if no value is provided with a metric name pairing.
| mp.metrics.distribution.histogram.buckets| Defines a custom set of (cumulative) histogram buckets for matching Histogram metrics to track and output.  Accepts a set of integer and decimal values for a metric name pairing.

> | mp.metrics.distribution.timer.buckets| Defines a custom set of (cumulative) histogram buckets for matching Timer metrics to track and output.  Accepts a set of decimal values with a time unit appended (that is, ms, s, m, h) for a metric name pairing.

Reviewer: ```suggestion
| mp.metrics.distribution.timer.buckets| Defines a custom set of (cumulative) histogram buckets for matching Timer metrics to track and output.  Accepts a set of decimal values with a time unit appended (`ms`, `s`, `m`, or `h`, for example) for a metric name pairing.
```

|mp.metrics.distribution.percentiles-histogram.enabled | Configures any matching Histogram or Timer metric to provide a large set of default histogram buckets to allow for percentile configuration with a monitoring tool. Accepts a true/false value for a metric name pairing.
| mp.metrics.distribution.histogram.max-value| When percentile-histogram is enabled for a Timer, this property defines an upper bound for the buckets reported. Accepts a single integer or decimal value for a metric name pairing.
| mp.metrics.distribution.histogram.min-value| When percentile-histogram is enabled for a Timer, this property defines a lower bound for the buckets reported. Accepts a single integer or decimal value for a metric name pairing.

> |mp.metrics.distribution.timer.max-value | When percentile-histogram is enabled for a Histogram, this property defines an upper bound for the buckets reported. Accepts a single decimal value with a time unit appended (that is, ms, s, m, h) for a metric name pairing. Accepts a single decimal value with a time unit appended (that is, ms, s, m, h) for a metric name pairing.

Reviewer: ```suggestion
|mp.metrics.distribution.timer.max-value | When percentile-histogram is enabled for a Histogram, this property defines an upper bound for the buckets reported. Accepts a single decimal value with a time unit appended (`ms`, `s`, `m`, or `h`, for example) for a metric name pairing.
```

> |mp.metrics.distribution.timer.min-value | When percentile-histogram is enabled for a Histogram, this property defines a lower bound for the buckets reported. Accepts a single decimal value with a time unit appended (that is, ms, s, m, h) for a metric name pairing.

Reviewer: ```suggestion
|mp.metrics.distribution.timer.min-value | When percentile-histogram is enabled for a Histogram, this property defines a lower bound for the buckets reported. Accepts a single decimal value with a time unit appended (`ms`, `s`, `m`, or `h`, for example) for a metric name pairing.
```
